source/vir/src/ast_to_sst.rs

```diff
@@ -161,11 +196,54 @@ fn function_can_be_exp(ctx: &Ctx, expr: &Expr, path: &Fun) -> Result<bool, VirEr
     }
 }
 
+pub fn assume_false(span: &Span) -> Stm {
+    let expx = ExpX::Const(Constant::Bool(false));
+    let exp = SpannedTyped::new(&span, &Arc::new(TypX::Bool), expx);
+    Spanned::new(span.clone(), StmX::Assume(exp))
+}
+
+/// Determine if it's possible for control flow to reach the statement after the loop exit.
+/// Naturally, we need to be conservative and answer 'yes' if we can't tell.
+/// However, this analysis is also relevant to the typing of the program: in particular,
+/// we ALSO need to return 'no' if any case where rustc's typechecker
+/// might have said 'no'.
+///
+/// The reason is that: if rustc determines that the loop can't exit, then
+/// the code after this will be unreachable, which means the user might be allowed
+/// to leave off a return expression. We need to detect that case, or else we might
+/// wrongly determine that it returns a 'unit' and we'll create malformed AIR code.
+///
+/// So when does Rust do this? As far as I can tell, it only does this if:
+///
+///   (i) it's a 'loop' statement, and
+///   (ii) it doesn't have ANY 'break' statement in it
+///        (It is possible that a 'break' statement in a while loop might itself be
+///        unreachable, but Rust doesn't seem to take that into account for this purpose.)
```

Comment:
Could we ask rustc for this information? I think this may be constructed for MIR, which would be potentially inconvenient for us, but it may be worth a look (e.g. for `break`, or for returns inside a loop). IIRC, the control flow graph is stored as https://docs.rs/rustc-cfg/0.4.0/rustc_cfg/struct.Cfg.html

Reply:
Ah, sorry, well I don't actually know that much about MIR, so I don't really know much what it would take to extract the information. The relevant control flow analysis, even with the breaks, on the other hand, probably isn't that hard (and we'll need that information anyway).

Yesterday I looked through rustc_typeck and found the place where it computes this, although I don't know how to get that information out either.